src/libaktualizr/bootloader/bootloader.cc

```diff
@@ -0,0 +1,94 @@
+#include "bootloader.h"
+#include "utilities/config_utils.h"
+#include "utilities/exceptions.h"
+#include "utilities/utils.h"
+
+std::ostream& operator<<(std::ostream& os, const RollbackMode mode) {
+  std::string mode_s;
+  switch (mode) {
+    case kUbootGeneric:
+      mode_s = "uboot_generic";
+      break;
+    case kUbootMasked:
+      mode_s = "uboot_masked";
+      break;
+    case kGrub:
+      mode_s = "grub";
+      break;
+    default:
+      mode_s = "none";
+      break;
+  }
+  os << '"' << mode_s << '"';
+  return os;
+}
+
+void BootloaderConfig::updateFromPropertyTree(const boost::property_tree::ptree& pt) {
+  std::string mode_s;
+  CopyFromConfig(mode_s, "rollback_mode", pt);
+  if (mode_s == "uboot_generic") {
+    rollback_mode = kUbootGeneric;
+  } else if (mode_s == "uboot_masked") {
+    rollback_mode = kUbootMasked;
+  } else if (mode_s == "grub") {
+    rollback_mode = kGrub;
+  } else {
+    rollback_mode = kBootloaderNone;
+  }
+}
+
+void BootloaderConfig::writeToStream(std::ostream& out_stream) const {
+  writeOption(out_stream, rollback_mode, "rollback_mode");
+}
+
+void Bootloader::setBootOK() const {
+  std::string sink;
+  switch (config_.rollback_mode) {
+    case kBootloaderNone:
+      break;
+    case kUbootGeneric:
+      if (Utils::shell("fw_setenv bootcount 0", &sink) != 0) {
+        LOG_WARNING << "Failed resetting bootcount";
+      }
+      break;
+    case kUbootMasked:
+      if (Utils::shell("fw_setenv bootcount 0", &sink) != 0) {
+        LOG_WARNING << "Failed resetting bootcount";
+      }
+      if (Utils::shell("fw_setenv upgrade_available 0", &sink) != 0) {
+        LOG_WARNING << "Failed resetting upgrade_available for u-boot";
+      }
+      break;
+    default:
```

Comment:
Shouldn't we have a `kGrub` case for `setBookOK` and `updateNotify`? Otherwise they will throw an exception no matter what if we try to use grub, which seems a bit harsh. It's true, they aren't implemented, but that shouldn't break functionality.

Reply:
Well, I see that there is nothing that sets the grub option, so maybe this doesn't really matter. Up to you. On the whole, this seems fine, but I'd like to test it or see a demo.

Reply:
Yes, maybe I should delete grub option at all to avoid confusion. Grub integration is not even scheduled yet.